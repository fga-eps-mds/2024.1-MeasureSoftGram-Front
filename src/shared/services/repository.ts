import axios, { AxiosRequestConfig, AxiosError } from 'axios';
import api from './api';
import { getAccessToken } from '@services/Auth';

interface RepositoryFormData {
  name: string;
  description?: string;
  url?: string;
  platform: string;
}

interface HistoricalCharacteristicsProps {
  organizationId?: string;
  productId?: string;
  repositoryId?: string;
  entity: string;
}

export type ResultSuccess<T> = { type: 'success'; value: T };
export type ResultError = { type: 'error'; error: Error | AxiosError | any };
export type Result<T> = ResultSuccess<T> | ResultError;
class Repository {
  private async getAuthHeaders(): Promise<AxiosRequestConfig['headers']> {
    try {
      const tokenResult = await getAccessToken();
      if (tokenResult.type === 'error' || !tokenResult.value.key) {
        return {};
      }



      return { Authorization: `Token ${tokenResult.value.key}` };
    } catch (error) {
      console.error('Failed to fetch authentication token:', error);
      return {};
    }
  }

  async createRepository(organizationId: string, productId: string, data: RepositoryFormData): Promise<Result<RepositoryFormData>> {
    try {
      const headers: AxiosRequestConfig['headers'] = await this.getAuthHeaders();
      if (!headers) {
        throw new Error('Access token not found.');
      }

      const response = await api.post(`/organizations/${organizationId}/products/${productId}/repositories/`, data, { headers });
      return { type: 'success', value: response?.data };
    } catch (err) {
      if (axios.isAxiosError(err)) {
        return { type: 'error', error: err };
      }
      return { type: 'error', error: new Error('Failed to create repository.') };
    }
  }


async updateRepository(organizationId: string, productId: string, repositoryId: string, data: RepositoryFormData): Promise<Result<RepositoryFormData>> {
  try {
    const headers: AxiosRequestConfig['headers'] = await this.getAuthHeaders();
    if (!headers) {
      throw new Error('Access token not found.');
    }

    const response = await api.put(`/organizations/${organizationId}/products/${productId}/repositories/${repositoryId}/`, data, { headers });
    return { type: 'success', value: response?.data };
  } catch (err) {
    if (axios.isAxiosError(err)) {
      return { type: 'error', error: err };
    }
    return { type: 'error', error: new Error('Failed to update repository.') };
  }
}


  async deleteRepository(organizationId: string, productId: string, repositoryId: string): Promise<Result<void>> {
    try {
      const headers: AxiosRequestConfig['headers'] = await this.getAuthHeaders();
      if (!headers) {
        throw new Error('Access token not found.');
      }
      const response = await api.delete(`/organizations/${organizationId}/products/${productId}/repositories/${repositoryId}/`, { headers });
      return { type: 'success', value: response?.data };
    } catch (err) {
      return { type: 'error', error: new Error('Failed to delete repository.') };
    }
  }

  async getHistoricalData(props: HistoricalCharacteristicsProps): Promise<Result<any>> {
    try {
      const { organizationId, entity, productId, repositoryId } = props;
      const headers: AxiosRequestConfig['headers'] = await this.getAuthHeaders();
      if (!headers) {
        throw new Error('Access token not found.');
      }
      const response = await api.get(`/organizations/${organizationId}/products/${productId}/repositories/${repositoryId}/historical-values/${entity}/`, { headers });
      return { type: 'success', value: response?.data };
    } catch (err) {
      return { type: 'error', error: new Error('Failed to fetch historical data.') };
    }
  }

  getRepository(organizationId: string, productId: string, repositoryId: string): Promise<Result<any>> {
    return api.get(`organizations/${organizationId}/products/${productId}/repositories/${repositoryId}/`);
  }

  getHistorical(props: HistoricalCharacteristicsProps): Promise<Result<any>> {
    const { organizationId, entity, productId, repositoryId } = props;
    return api.get(
      `organizations/${organizationId}` +
        `/products/${productId}/repositories/${repositoryId}` +
        `/historical-values/${entity}/`
    );
  }

  getLatest(props: HistoricalCharacteristicsProps) {
    const { organizationId, entity, productId, repositoryId } = props;
    return api.get(
      `organizations/${organizationId}` +
        `/products/${productId}/repositories/${repositoryId}` +
        `/latest-values/${entity}/`
    );
  }

  getTsqmiBadgeUrl(props: HistoricalCharacteristicsProps) {
    const { organizationId, entity, productId, repositoryId } = props;
    return (
      api.getUri() +
      `/organizations/${organizationId}` +
      `/products/${productId}/repositories/${repositoryId}` +
      `/latest-values/${entity}/badge`
    );
  }
}

export const repository = new Repository();
Object.freeze(repository);
