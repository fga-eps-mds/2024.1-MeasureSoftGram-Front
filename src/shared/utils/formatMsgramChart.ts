import { Historical } from '@customTypes/repository';
import { format } from 'date-fns';
import _ from 'lodash';
import convertToCsv from './convertToCsv';

interface Props {
  historical: Historical[];
  title: string;
  isEmpty: boolean;
}

const formatMsgramChart = ({ historical, title, isEmpty = false }: Props) => {
  const legendData = _.map(historical, 'name');
  const historicalData = _.map(historical, 'history');
  const dates = _.uniq(historicalData.flat(1).map((h) => format(new Date(h.created_at), 'dd/MM/yyyy HH:mm')));
  const xAxisData = dates.flatMap((date) => [date, '                ']);

  const numberOfGraphs = legendData?.length ?? 0;

  const grid = _.times(numberOfGraphs, (i) => ({
    show: false,
    left: '120px',
    right: '4%',
    height: '60px',
    y: `${60 * i + 60}px`,
    containLabel: true
  }));

  const legend = _.times(numberOfGraphs, (i) => ({
    show: !isEmpty,
    selectedMode: false,
    x: 0,
    y: `${60 * i + 69}px`,
    data: [legendData?.[i] ?? '-']
  }));

  const xAxis = _.times(numberOfGraphs, (i) => ({
    show: i === numberOfGraphs - 1 && !isEmpty,
    gridIndex: i,
    type: 'category',
    data: xAxisData
  }));

  const yAxis = _.times(numberOfGraphs, (i) => ({
    show: !isEmpty,
    max: 1,
    min: 0,
    type: 'value',
    gridIndex: i,
    splitNumber: 1,
    axisLabel: {
      show: false
    },
    splitLine: {
      show: false,
      lineStyle: {
        width: 2
      }
    }
  }));

  const series = historical?.map((h, i) => ({
    show: !isEmpty,
    name: h.name,
    type: 'line',
    data: h.history.flatMap(({ value }, index) => {
      const formattedValue = { value: [index * 2, value.toFixed(3)] };
      const zeroPoint = { value: [index * 2 + 1, 0], symbol: 'none' };
      return index < h.history.length - 1 ? [formattedValue, zeroPoint] : [formattedValue];
    }),
    xAxisIndex: i,
    yAxisIndex: i,
    showSymbol: true
  }));

  const dataZoom = [
    {
      show: !isEmpty,
      type: 'slider',
      y: `${60 * numberOfGraphs + 60}px`,
      xAxisIndex: _.times(numberOfGraphs, (i) => i)
    }
  ];

  const handleExportCsv = () => {
    if (historical) {
      const csvContent = convertToCsv(historical);

      const blob = new Blob([csvContent], { type: 'text/csv' });
      const url = window.URL.createObjectURL(blob);
      const a = document.createElement('a');
      a.href = url;
      const timestamp = new Date().toISOString().replace(/[. ]/g, '');
      a.download = `msgram-${title}-${timestamp}.csv`;
      a.click();
      window.URL.revokeObjectURL(url);
    }
  };

  return {
    title: {
      text: title
    },
    tooltip: {
      show: !isEmpty,
      trigger: 'item'
    },
    toolbox: {
      feature: {
        myCustomTool: {
          show: true,
          title: 'Export CSV',
          icon: 'image:///images/png/iconCsv.png',
          onclick: () => {
            handleExportCsv();
          }
        }
      }
    },
    dataZoom,
    grid,
    xAxis,
    yAxis,
    series,
    legend
  };
};

export default formatMsgramChart;
