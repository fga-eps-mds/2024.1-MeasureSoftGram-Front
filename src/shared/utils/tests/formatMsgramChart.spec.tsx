import { Historical } from '@customTypes/repository';
import formatMsgramChart from '../formatMsgramChart';

describe('formatMsgramChart', () => {
  const chartTitle: string = 'title-test';
  const dateNow = new Date(1701132814785);

  it('should return the correct chart configuration with non-empty historical data', () => {
    const historical: Historical[] = [
      {
        id: 1,
        key: 'reliability',
        name: 'Reliability',
        history: [
          {
            id: 11,
            value: 10,
            created_at: dateNow,
          },
          {
            id: 12,
            value: 20,
            created_at: dateNow,
          },
        ],
        latest: {
          id: 12,
          value: 32,
          created_at: dateNow,
        },
      },
      {
        id: 2,
        key: 'maintainability',
        name: 'Maintainability',
        history: [
          {
            id: 21,
            value: 30,
            created_at: dateNow,
          },
          {
            id: 22,
            value: 40,
            created_at: dateNow,
          },
        ],
        latest: {
          id: 22,
          value: 50,
          created_at: dateNow,
        },
      },
    ];

    const params = {
      historical,
      title: chartTitle,
      isEmpty: false,
    };

    const chartConfig = formatMsgramChart(params);

    expect(chartConfig).toMatchSnapshot();
  });

  it('should return the correct chart configuration with empty historical data', () => {
    const historical: Historical[] = [];

    const params = {
      historical,
      title: chartTitle,
      isEmpty: true,
    };

    const chartConfig = formatMsgramChart(params);

    expect(chartConfig).toMatchSnapshot();
  });
});
