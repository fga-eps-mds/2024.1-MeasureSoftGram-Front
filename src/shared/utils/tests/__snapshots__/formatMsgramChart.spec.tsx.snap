// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatMsgramChart should return the correct chart configuration with empty historical data 1`] = `
Object {
  "dataZoom": Array [
    Object {
      "show": false,
      "type": "slider",
      "xAxisIndex": Array [],
      "y": "60px",
    },
  ],
  "grid": Array [],
  "legend": Array [],
  "onEvents": Object {
    "datazoom": [Function],
  },
  "series": Array [],
  "title": Object {
    "text": "title-test",
  },
  "toolbox": Object {
    "feature": Object {
      "myCustomTool": Object {
        "icon": "image:///images/png/iconCsv.png",
        "onclick": [Function],
        "show": true,
        "title": "Export CSV",
      },
    },
  },
  "tooltip": Object {
    "show": false,
    "trigger": "axis",
  },
  "xAxis": Array [],
  "yAxis": Array [],
}
`;

exports[`formatMsgramChart should return the correct chart configuration with non-empty historical data 1`] = `
Object {
  "dataZoom": Array [
    Object {
      "show": true,
      "type": "slider",
      "xAxisIndex": Array [
        0,
        1,
      ],
      "y": "180px",
    },
  ],
  "grid": Array [
    Object {
      "containLabel": true,
      "height": "60px",
      "left": "120px",
      "right": "4%",
      "show": false,
      "y": "60px",
    },
    Object {
      "containLabel": true,
      "height": "60px",
      "left": "120px",
      "right": "4%",
      "show": false,
      "y": "120px",
    },
  ],
  "legend": Array [
    Object {
      "data": Array [
        "Reliability",
      ],
      "selectedMode": false,
      "show": true,
      "x": 0,
      "y": "69px",
    },
    Object {
      "data": Array [
        "Maintainability",
      ],
      "selectedMode": false,
      "show": true,
      "x": 0,
      "y": "129px",
    },
  ],
  "onEvents": Object {
    "datazoom": [Function],
  },
  "series": Array [
    Object {
      "data": Array [
        Object {
          "value": "10.000",
        },
        Object {
          "value": "20.000",
        },
      ],
      "name": "Reliability",
      "show": true,
      "type": "line",
      "xAxisIndex": 0,
      "yAxisIndex": 0,
    },
    Object {
      "data": Array [
        Object {
          "value": "30.000",
        },
        Object {
          "value": "40.000",
        },
      ],
      "name": "Maintainability",
      "show": true,
      "type": "line",
      "xAxisIndex": 1,
      "yAxisIndex": 1,
    },
  ],
  "title": Object {
    "text": "title-test",
  },
  "toolbox": Object {
    "feature": Object {
      "myCustomTool": Object {
        "icon": "image:///images/png/iconCsv.png",
        "onclick": [Function],
        "show": true,
        "title": "Export CSV",
      },
    },
  },
  "tooltip": Object {
    "show": true,
    "trigger": "axis",
  },
  "xAxis": Array [
    Object {
      "data": Array [
        "27/11/2023 21:53",
      ],
      "gridIndex": 0,
      "show": false,
      "type": "category",
    },
    Object {
      "data": Array [
        "27/11/2023 21:53",
      ],
      "gridIndex": 1,
      "show": true,
      "type": "category",
    },
  ],
  "yAxis": Array [
    Object {
      "gridIndex": 0,
      "max": 1,
      "min": 0,
      "show": true,
      "splitLine": Object {
        "lineStyle": Object {
          "width": 2,
        },
      },
      "splitNumber": 1,
      "type": "value",
    },
    Object {
      "gridIndex": 1,
      "max": 1,
      "min": 0,
      "show": true,
      "splitLine": Object {
        "lineStyle": Object {
          "width": 2,
        },
      },
      "splitNumber": 1,
      "type": "value",
    },
  ],
}
`;
